### ISSUE TITLE
Zoom and default-zoom

### ISSUE THREAD

**Issue by zerodogg** [sonnyp/Tangram#79]
My laptop has a semi-hidpi screen. The default font size in Tangram is too small. I have set systemwide scaling a bit up, but not enough for tangram to become quite usable. In firefox I've got it set to default to 120% zoom.

It would be useful if Tangram responded to the usual zoom hotkeys (ctrl+, ctrl-, and ctrl-mousewheel), and had a way to set the default zoom level. Alternatively being able to inject css would work as well as a stop-gap.

**Comment by sonnyp** [comment]
> I have set systemwide scaling a bit up, but not enough for tangram to become quite usable

Does it mean Tangram doesn't behave the same as other applications on your system? Can you be more specific about your system and what kind of scaling you applied?

> In firefox I've got it set to default to 120% zoom.

So I get your scaling doesn't work for Firefox?

> It would be useful if Tangram responded to the usual zoom hotkeys (ctrl+, ctrl-, and ctrl-mousewheel), 

Agree (and ctrl0 to reset). GNOME Web supports ctrl+/ctrl-/ctrl0 so it should be trivial. ctrl+mousewheel - not sure. 

Would you like to give it a try? I'm happy to give some pointers.

**Comment by psydvl** [comment]
I'm using [AUR version](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=tangram) with KDE on Manjaro.
Tangram have another scale than Firefox/Chrome.
I have Global scale 125% what have effect on all apps, but looks like not for Tangram.
<kbd>Ctrl</kbd><kbd>+</kbd>/<kbd>Ctrl</kbd><kbd>-</kbd> not work; only for developer console scale(<kbd>Ctrl</kbd><kbd>shift</kbd><kbd>I</kbd>), not for service page. <kbd>Ctrl</kbd>+<kbd>mousewheel</kbd> not works at all

**Comment by sonnyp** [comment]
On Fedora 32; Tangram responds to GNOME/Wayland scaling.

Can you reproduce the issue with [GNOME Web flatpak](https://flathub.org/apps/details/org.gnome.Epiphany)?

**Comment by zerodogg** [comment]
This isn't actually a bug, I should perhaps have been clearer. It responds to scaling, it's just that I can't put systemwide scaling as high as Tangram would need it to be, because then gnome-shell's top bar becomes much larger than I'd like. Thus, having a separate zoom inside of Tangram would be a solution (and it is the solution I've used in firefox).

**Comment by sonnyp** [comment]
So I learned a bit more about this.

There are 2 scaling factors in GNOME / Mutter

The first one is scaling factor `Settings -> Displays -> Scale` to which Tangram should respond well.

The second one is the font scaling factor; `Settings -> Accessibility -> Large Text` or `GNOME Tweaks -> Fonts -> Scaling Factor`

According to https://webkitgtk.org/2021/01/12/webkitgtk2.31.1-released.html - WebkitGTK 2.32 will implement the second which would add support in Tangram as well. However Epiphany 3.38.2 with WebkitGTK 2.30.4 already seems to respond to font scaling factor well so I'm not sure what this is about - I will look into it if needed.

@psydvl can you please open a separate ticket? It looks like unrelated

@zerodogg would font scaling factor solve your issue?

I'm not against having zoom control in Tangram but I don't really have a need for it myself so perhaps we can find an other solution.

However if you'd like to give it a try - I'd be happy to help and provide guidance.

**Comment by PolGZ** [comment]
Debian 10 user. Here in "oldkerneland", we don't have fractional scaling, only 100-200% dpi, so I'm not willing to loose half of my screen on menus, haha.

The second two options (`Settings -> Accessibility -> Large Text` and `GNOME Tweaks` doesn't seem to work on Tangram (flatpak) content. Nor I think that would be a viable option, because I find the general appearance of the system quite big and ugly looking, IMHO.

**Comment by sonnyp** [comment]
Thanks for the feedback

> Debian 10 user. Here in "oldkerneland", we don't have fractional scaling, only 100-200% dpi, so I'm not willing to loose half of my screen on menus, haha.

In case you are interested, here are the instructions to enable fractional scaling - yes even for oldkerneland :)

https://wiki.archlinux.org/index.php/HiDPI#Fractional_scaling 

https://www.linuxuprising.com/2019/04/how-to-enable-hidpi-fractional-scaling.html

> The second two options (Settings -> Accessibility -> Large Text and GNOME Tweaks doesn't seem to work on Tangram (flatpak) content.

Strange - works for me but ok if that's not what you want it doesn't matter I guess.

Fair - I will consider implement Zoom in/out then if nothing else works out

Meanwhile, if anyone is interested in giving it a try I'm happy to assist - should be fairly easy if you have a minimum knowledge of JavaScript.

**Comment by PolGZ** [comment]
> In case you are interested, here are the instructions to enable fractional scaling - yes even for oldkerneland :)
> https://wiki.archlinux.org/index.php/HiDPI#Fractional_scaling

I didn't know about that possibility! Thanks to point it out. But, unfortunately, at least under Wayland, the result is too bad! Everything is very pixelated and blurry. I haven't tried on X11 (because the integration of the megasync app on Gnome X11 is really annoying (it doesn't minimize to tray icon only) and I have it on autostart, so I'm stuck in Wayland).

> Strange - works for me but ok if that's not what you want it doesn't matter I guess.

I have checked that the behaviour for text/fonts options is the same under X11, in case it was something of Wayland. I don't know what different.

> Fair - I will consider implement Zoom in/out then if nothing else works out

For now, it seems that the rest of the options don't work with the flatpak version under Debian 10 (Gnome 3.30.2). That would be very much appreciated.

**Comment by bear454** [comment]
I'm in just about the same boat as @zerodogg ... running browsers at 120% zoom, to get text size to match the rest of the desktop.

I added an ENV var to the tangram .desktop file, so I'm running `GDK_DPI_SCALE=1.2 flatpak run re.sonny.Tangram` ... which accomplishes the same thing.

While it would be nice to have the zoom controls match what's implemented in gnome-web, I'm happy I've got a way to get the size right (and won't be futzing with it much).

**Comment by guss77** [comment]
I'm running on KDE with a QHD display without scaling, and while both Firefox and Chrome look fine, both Tangram and Epiphany have tiny fonts to the point that it is almost unreadable. The rest of the UI looks fine.

> I added an ENV var to the tangram .desktop file, so I'm running `flatpak run --env=GDK_DPI_SCALE=1.2 re.sonny.Tangram` ... which accomplishes the same thing.

That didn't do anything for me.

![tangram-hangouts](https://user-images.githubusercontent.com/381782/136943531-06aaaf08-f7fb-4da0-b793-37485bb321fd.png)

As can be seen in the screenshot, the font-size is reported as 9px. I think it is not even 9px, but interestingly the CSS shows it should be 24px and also changing the font size manually in the inspected doesn't affect the computed size...

**Comment by jiri-one** [comment]
Yes, I am not able to use Tangram, I always need to adjust size of fonts in opened pages - Its not problem in Ferdi, Rambox, Franz or on Epiphany (like Web on Gnome) which I am using in app mode for this now.

**Comment by forteller** [comment]
I came to Github after trying v2.0 for a few minutes, just to request this feature. I take the chance on commenting, since I see it's been a while without any developments here. Thanks for a great tool, which would be even greater if it was easy to make the text more readable for everyone, no matter their screen resolution or eyesight :)

**Comment by bennypowers** [comment]
This is a critical accessibility issue. Every browser should allow users to scale their own fonts. Please put "GNOME design principles" or whatever aside and put this issue on the roadmap. Every other browser has this.

**Comment by sonnyp** [comment]
> put "GNOME design principles" or whatever aside

What are you talking about? Did you even read the discussion?

> put this issue on the roadmap

If you had made the effort of reading the discussion instead of making demands, you would have seen that I already agreed to the feature request in the first comment.

This is free software, I don't owe you anything and since you happen to be a developer:

Please put the feature on your to-do list and get it done.

**Comment by RushingAlien** [comment]
Hello,

I think what people are saying is zoom in in the style if you press `ctrl`, `+` in browsers. This is treu for the likes of Electron apps and Ferdium, also browsers like Firefox and Chromium

[Screencast dari 2023-09-29 13-39-22.webm](https://github.com/sonnyp/Tangram/assets/80886145/b379dca7-c50d-47fd-9993-252522580ffd)
